### PR TITLE
Bilateral Fischer-Burmeister constraints and a Maliar method user guide

### PR DESCRIPTION
**Pull request by alanlujan91**
## Summary

Adds bilateral inequality-constraint support to the Euler loss and a user guide for the Maliar method. Closes #215.

## Constraints

- `Control` accepts numeric bounds directly. A number becomes a zero-argument callable at the `Control` boundary through `normalize_bound`, and a boolean raises a `TypeError` so `upper_bound=True` cannot slip through as `1.0`. Existing callable bounds are unchanged.
- `EulerEquationLoss(constrained=True)` now covers both bounds with a bilateral Fischer-Burmeister residual, `FB(s_u, -FB(s_l, -f))`. When one bound is slack the residual collapses to the one-sided form, so a model with only an upper bound behaves exactly as before.
- The constraints user guide is rewritten for the bilateral loss and the numeric-bound API, with a table showing where each mechanism (network transform, complementarity loss, VBI box constraints) applies.

## Maliar method user guide

- New `docs/user_guide/maliar.md` explains the all-in-one expectation operator (why multiplying two independent residual draws yields the unbiased estimator), the Euler and Bellman losses and how they differ on slope-versus-level identification, `maliar_training_loop`, and the link to constrained controls.
- The Algorithms guide links to the new page. A leftover duplicate "Solving a block directly" heading there is retitled "Overview".

## Testing

- Test suite: 387 passed, 3 skipped (the slow RL-convergence tests).
- Docs build clean, with no broken cross-references.

Drafted with assistance from Claude Code (Opus 4.8).

### CODE REVIEW

File: docs/user_guide/maliar.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+optimality condition the residual $f$ encodes.
+
+### The Euler Loss
+
+The Euler loss, {py:class}`~skagent.loss.EulerEquationLoss`, encodes the
+first-order condition. For a consumption-saving model it equates the marginal
+utility of consuming today with the discounted expected marginal utility of
+consuming tomorrow. Let $u$ be the utility function, $c_t$ consumption in period
+$t$, $\beta$ the discount factor (read from the model's discount variable), and
+$R$ the gross return on savings. The residual is
```
Comment (sbenthall) [inline docs/user_guide/maliar.md]: This wording is specific to the consumption model.

Consider using the framing from the original MMW paper, which was more genera.

File: docs/user_guide/maliar.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+Many dynamic programs have no closed-form solution, and discretizing the state
+space scales badly as states accumulate. The Maliar method (Maliar, Maliar, and
+Winant 2021) sidesteps both problems: it represents the decision rule as a
+neural network and turns the search for a policy into minimizing an
+equation-residual loss over a sample of states. This page explains what the
+method minimizes, the two residual losses scikit-agent provides, and how to run
+the training loop.
+
+## The All-in-One Expectation Operator
```
Comment (sbenthall) [inline docs/user_guide/maliar.md:27]: I think this section would be daunting for a software developer with no background to read. Is there a way to make it easier?

File: docs/user_guide/maliar.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+a policy head and a value head that one optimizer trains together.
+
+### Choosing Between Them
+
+| Property                  | Euler loss          | Bellman loss          |
+| ------------------------- | ------------------- | --------------------- |
+| Optimality condition      | first-order (Euler) | Bellman equation      |
+| Requires a value function | no                  | yes                   |
+| Pairs with                | `BlockPolicyNet`    | `BlockPolicyValueNet` |
+| Identifies                | slope only          | slope and level       |
```
Comment (sbenthall) [inline docs/user_guide/maliar.md:148]: This is cool

File: docs/user_guide/maliar.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import skagent.loss as loss
+
+euler_loss = loss.EulerEquationLoss(bp, parameters=calibration, constrained=True)
+
+policy, states = maliar.maliar_training_loop(
+    bp,
+    euler_loss,
+    states_0,
+    calibration,
+    shock_copies=2,  # independent draws for the all-in-one operator
```
Comment (sbenthall) [inline docs/user_guide/maliar.md:174]: How do you decide how to set this?

File: docs/user_guide/maliar.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    max_iterations=40,
+    tolerance=1e-6,
+    network_width=32,
+)
+```
+
+Resampling is what makes this more than gradient descent on a fixed grid:
+trained only on the initial grid, the network would learn the policy where the
+grid happens to sit, not where the model actually spends its time. The loop
+returns the trained policy network and the final panel of training states.
```
Comment (sbenthall) [inline docs/user_guide/maliar.md]: To me, this is the big win of the Maliar method, and could be foregrounded at the beginning of the user guide.

The Maliar method uses neural networks and resampling to beat the curse of dimensionality that appears in grid methods. Also, unlike PPO, it uses the model directly rather than being model-free. These are the big benefits of this method.

File: src/skagent/block.py [hunk 6]
```diff
@@ -40,20 +84,24 @@ class Control:
...
     agent : str
         A label identifying the agent role to which this control is attributed.
     """
 
     def __init__(self, iset, lower_bound=None, upper_bound=None, agent=None):
         self.iset = iset
-        self.lower_bound = lower_bound
-        self.upper_bound = upper_bound
+        self.lower_bound = normalize_bound(lower_bound, "lower_bound")
+        self.upper_bound = normalize_bound(upper_bound, "upper_bound")
```
Comment (sbenthall) [inline src/skagent/block.py:104]: this is a cool way to solve this problem

File: tests/test_torch.py [hunk 7]
```diff
@@ -16,7 +16,7 @@
     # },
     "dynamics": {
         "m": lambda a, q: a * q,
-        "c": Control(["m"], upper_bound="a"),
+        "c": Control(["m"]),
```
Comment (sbenthall) [inline tests/test_torch.py:19]: Here, there was an idea that the bound could be a named symbol or maybe a string expression, which I suppose would be turning into a lambda somewhere along the way.

This seems to be removing this case, which is fine.

But maybe make a new issue for this if it's not something you want to cover in this PR?

If you want to introduce HABLO in v.0.2. then this seems to be aligned with having more parseable math expressions used instead of explicit lambdas.
